### DISCUSSION TITLE
Handling multiple representations

### DISCUSSION THREAD

**Discussion by hepcat72** [Princeton-LSI-ResearchComputing/tracebase#1191] [General]
Michael's recent study submission with labeled Oxygen in water has numerous instances of "multiple representations" of compounds in peak annotation files whose samples are derived from the same sequence.

When we were faced with this issue and were handling mzXML loading, we made a choice to not support multiple representations of compounds, meaning that the researcher must decide which of multiple representations of each compound to keep and which one to discard. I think the thinking was that the number of cases to deal with would be small. However, in the case of Michael's recent study, that does not appear to be true.

One goal we had was to smooth the loading process and remove barriers to submission, but it seems that handling multiple representations in this way is such a barrier we have added.

Michael had responded to my summary of the load attempt before I left on vacation:

> Thanks for summarizing this before your vacation!
> 
> The water_errors.txt file is, like you said, pretty overwhelming...so many instances of multiple representations! I know that many of these are the same compound curated from two different scans (usually a negative mode and a positive mode).
> 
> I don't have access to tracebase-staging, or at least I don't know where to find it.
> 
> In general, I'm worried about how we're handling this problem. As we've discussed, the benefit is that users who are analyzing data in tracebase do not have to worry about multiple representations, which I do not want to deviate from. The 2 main issues are
> 
> - it's very labor intensive (happens in many submissions, and takes much longer to fix than other frequent problems like Compound synonyms)
> - we have users manually deleting data from the sheets they submit, then uploading updated versions of that file.
> Could we 1) avoid altering the original data file directly and 2) provide an interface or method for the user to quickly see the multiple representations and select their preferred option? I'm picturing something like this:
> image
> 
> This way, we are not deleting the original data, and we can change this later if choices change. Of course, Rob's script to summarize the validation errors could be an important step towards something like this.

**Comment by hepcat72** [comment]
@mneinast

Regarding the adenine error, yes, I can make an edit in the load script to treat that like a warning and simply skip that peak group load.

WRT multiple representations, those errors are overwhelming because you're getting an error for every sample column. Ideally, I would have a summarizing error which just lists the common compounds between 2 files. (There should exist an error making this summarization, but apparently there is a minor bug that is preventing it from happening.)

I think that, given the goal of removing barriers to submission, the decision we made to not allow multiple representations turns out to have been counter to that goal. Regardless of whether we make it easier to choose which compound representation to keep and which to discard, it still prevents/slows loading.

In the past, the way we handled this was to add discriminating information to split up the MSRunSample record into multiple records. E.g. polarity. We still have a way to discriminate between records (the mzXML file), but we had decided 2 things:

1. we didn't want to have to have the mzXML files ready before loading because it takes time to tease apart which goes with which peak annotation file
2. We didn't want the confusion of multiple measurements for a compound in a sample in a sequence

The problem in the schema is that we're using placeholder MSRunSample records (before the mzXMLs are loaded) and a PeakGroup's name (and linked MSRunSample record) must be unique. When we get to the second compound representation, we get a conflicting value error because the peak annotation file associated with the peak group has a different conflicting file associated with the "unique" peak group name / compound.

I have an idea for a quick(ish) fix. The problem with multiple representations is one of transparency. When you have multiple valine measurements, you don't know what's going on with it. If we had a way to contextualize the measurement, then having multiple representations maybe isn't such a big deal. Users would know what they are looking at and could consume the information with an understanding of what it is they're looking at. So we could append something to the peak group name to provide that context - or another field. And if we want users to be able to select their preferred measurement, we could provide a way to do that too, like the mock-up you provided, but without omitting loaded data.

**Reply by hepcat72** [reply]
Incidentally, we're in the car RN. almost to our destination.

Another option is to mark peak groups with multiple representations with like a footnote whenever they're presented that is explained in a tooltip. If associated with an mzXML, the polarity and scan range could be provided.

**Reply by hepcat72** [reply]
Copying my message about Adenine from slack:

On second leg of our trip...

Can I ask what the utility is of loading a peak group that doesn't have any elements that are labeled in the tracers?

Right now, there is code that likely assumes that every measured compound has an element that *could* be labeled.

Let me venture a guess... the biochemistry that results in a labeled element ending up in another compound has byproducts that may not have any elements that are labeled. Quantitatively measuring those unlabeled byproducts could lead to insights into how the labeled element gets into the resulting (other) measured compound? I.e. one chemical pathway versus another?

If that's the case, that would make sense. I would have to check over the code to see what problems there could be related to measured compounds without labeled elements... could you create an issue that describes this?

**Reply by mneinast** [reply]
adding my reply:  why capture a compound without the labeled element?

Basically, it is useful if the researcher needs to check total ion counts.  So this is not directly within the scope of tracebase (focused mainly on fractional labeling), but it is an ancillary objective that would be useful to implement.  If it is an easy change (ie one day to draft) then I think it's worth adding at some point.  But not before milestone 3.0.

**Comment by hepcat72** [comment]
OK @mneinast - I just created an example of how I could represent the selections that need to be made for the conflicting compounds ("multiple peak group representations"). I created drop-down examples for the first 30 rows.

Researchers would be required to pick a file in the dropdowns for every row.

Note, this is a rough mock-up.  Some rows are missing sequence names (but the code could easily include them). The sequence name elements are not in the same order as they are represented in other sheets (which is due to the way I quickly created this sheet).

A few notes:
- The "Conflicting Compound" is the peak group name, so it won't necessarily correspond directly to a compound in the compounds sheet.  Perhaps I should name it "Peak Group Name" instead, to avoid confusion?
- The sequence name is necessary to be able to identify the files in which the compound must be skipped.
- Compounds repeat because they are multiply represented in multiple sequences' peak annotation files.
- The name of the added sheet is "Compound Measurement Conflicts".  I only ever expect to include it IF there are multiple representations.
- Alternatively, I could name the sheet "Measured Compounds" (or "Peak Groups") and include every measured compound and only have the dropdowns for the rows where there is a conflict. *(However, now that I think about it, that would be an enormous sheet - perhaps overwhelming, especially given how many non-conflicting files [from different sequences] can have the same compound, just with different samples.)*

[MASTER STUDY DOC FOR TRACEBASE_exp011b3_240816-edit-mock.xlsx](https://github.com/user-attachments/files/16881012/MASTER.STUDY.DOC.FOR.TRACEBASE_exp011b3_240816-edit-mock.xlsx)

**Comment by hepcat72** [comment]
I am linking issue #1201 here.  It contains relevant information to this topic, namely:

- A demonstration of data in the database that shows we have sequences extant in the DB that ran the same set of samples multiple times (i.e. they're not time points).
- The code that prefers the arbitrarily "last" sequence sample's peak group.

**Reply by lparsons** [reply]
I was looking for some additional info on each multiple representation, so I reworked the query. It's not nearly as concise a solution as yours, @hepcat72, but it does output the peak annotation files and the study names. Also, for some reason I get 202 records, not 201 like you did. I'm not sure what is different.

The script is at `tracebase.princeton.edu:/var/www/tracebase/tracebase-multiple-representations.py`.
```python
#!/usr/bin/env python
# coding: utf-8

from DataRepo.models import *
import pandas as pd
import os

pgdf = pd.DataFrame.from_dict(PeakGroup.objects.values('id', 'msrun_sample__sample', 'compounds'))

sample_compound_count = pgdf.groupby(['msrun_sample__sample', 'compounds']).count()
multiple_representations = sample_compound_count.loc[(sample_compound_count['id'] > 1)]
multiple_representations.reset_index()

print("sample\tcompound\tpeak_annotation_files\tstudies")
for row in multiple_representations.reset_index().itertuples():
    sample = Sample.objects.get(pk=row.msrun_sample__sample)
    compound = Compound.objects.get(pk=row.compounds)
    studies = list(sample.animal.studies.all().values_list("name", flat=True))
    peak_groups = PeakGroup.objects.filter(msrun_sample__sample=sample, compounds=compound)
    print(f"{sample}\t{compound}\t{list(peak_groups.values_list('peak_annotation_file__filename', flat=True))}\t{studies}")
```

The results are in a Google Sheet: https://docs.google.com/spreadsheets/d/1HsLdBP1AU4OqTphWhUEtRn6lGTzp-gkBAe2Ui5rHncY/edit?usp=sharing

**Comment by lparsons** [comment]
@hepcat72 brought up an excellent point about the subtle difference between a Peak Group name and the compound links to a peak group: https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/1201#issuecomment-2356049640.

In my view, the peak group name has been a "note" that records what the researcher recorded during peak picking, but the compound links make an explicit deterministic annotation of a peak group to specific compound(s).

**Reply by hepcat72** [reply]
I think we've all made assumptions about the peak group name, or at least we didn't document its intended usage well and I simply forgot. If it is just a note, then we probably should use something other than it in the unique constraints:
```
        constraints = [
            models.UniqueConstraint(
                fields=["name", "msrun_sample"],
                name="unique_peakgroup",
            ),
        ]
```
To summarize my comment here, my assumption had been that if we record a name using a compound synonym (instead of the primary compound name), specifically to support differentiation of a qualitatively different compound using the synonym (e.g. stereo-isomer), then it seems logical to support multiple different stereo-isomer peak groups linked to the same compound.

If we were to use the linked compounds in the unique constraint somehow (which I'm not sure is possible given the M:M relationship), then that would prohibit multiple peak groups of the same "compound" linked to different stereo-isomers.

Michael has indicated to me (in related discussions with me) that he's not sure we should continue to support the differentiation of stereo-isomers.

**Comment by hepcat72** [comment]
Meeting 9/18/2024 decisions/notes:

- Base PeakGroup unique constraint on compound(s) and sample (not on name, sample, and sequence)
- A PeakGroup linked to citrate, and one linked to both citrate and isocitrate, will be "allowed" with a warning
- Peak Group name is an "annotation"/"note"

**Comment by hepcat72** [comment]
Lance pushed back a little on my PR (which he approved).  I suspect that he feels that the hidden "Common Samples" column may not need to exist and we went back and forth on it a little bit [in the comments](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/1209#discussion_r1769320493).  I thought that we should move that discussion here.

## The `Peak Group Conflicts` sheet/file

First, let me reiterate a couple clarifications I made in that back and forth:

1. The "Example Samples" column *appears* to be a duplicate of the "Common Samples" column, but it is not.  The example column currently contains 1 sample per row, though the test file has 3 (because after I created the test file, I decided 3 looked too busy).
2. The `Peak Group Conflicts` sheet contains 5 columns in total:
   - `Peak Group Conflict` (a lower-cased sorted peak group name [to be case and order insensitive])
   - `Selected Peak Annotation File` (the dropdown for selecting the file from which the selected representation should come from)
   - `Common Sample Count` (convenience data for informational purposes only)
   - `Example Samples` (convenience data for informational purposes only containing the first sample from the next [hidden] column...)
   - `Common Samples` (a **hidden** column containing `;`-delimited DB sample names)
3. The researcher must select a file from which the selected peak group should be loaded.  All other versions of that peak group in other files is skipped.  I will refer to this as a "conflict resolution", and it's notable to also point out below that I will refer to (and explain) a "conflict resolution conflict".

## Purpose

The common samples column was created to serve 2 purposes:

1. It serves as a means to catch "conflict resolution conflicts".
   - *Conflict resolution conflicts* would result from manual manipulation of the data in the sheet (e.g. the user copies conflict resolutions from one in-progress study doc to another, and in a duplicate row, selects different files to resolve the same conflict).
   - This is preemptively "caught" using a unique constraint applied to the file.  The code can see the combination of the compound and the common samples to say "hey, you've got a duplicate row here and you selected different files to resolve that same conflict - you need to delete one".
2. Having conflict resolutions tied to handle cases where the files involved do not all have the same identical sets of samples.
   - It allows us to not inappropriately skip peak groups (when the sample set in the overlapping files are not entirely the same), for example:
     - Files (containing e.g., `Lysine`) involved:
       - File 1 has 20 samples
       - File 2 has 25 samples (where 20 are in common with file 1)
     - User selects file 1 as the selected representation
     - Without access to the samples, we would skip `Lysine` for all 25 samples from file 2
     - With the samples, we can load `Lysine` from those extra 5 samples, because we know they don't have multiple representations.
   - It allows us to select different resolutions for multiple subsets of samples that overlap in a different set of files from another set of files.  For example (going with the above example):
     - Files (containing e.g., `Lysine`) involved:
       - File 1 has 20 samples
       - File 2 has 25 samples (where 20 are in common with file 1)
       - File 3 has 5 samples (all in common with file 2 and not in common with file 1)
     - User selects:
       - File 1 as the selected representation for the 20 samples
       - File 3 as the selected representation for the 5 samples
     - File 3 could represent a "remainder" of the 5 samples that they forgot to include in the run that produced file 1
   - I could be wrong, but I suspect that neither of these cases are likely to exist in the data we have thus far, but as Michael has pointed out, these situations are possible.

## Notes

- If a researcher makes multiple submissions on the same study, it is very likely that we will end up in a situation where there exists a "multiple representation" where a peak group was previously loaded, but it's the as-yet to load peak group should be the selected representation.  This means that the loader will have to be able to replace 1 peak group version with another.  It could even come from a different sequence.  So the resolution in this case would mean that the loader must be able to delete a previously loaded record.
- I intend to use the "Common Samples" column in the future to help users figure out "conflict resolution conflicts" by revealing that column when such an error occurs.

## Discussion

It is possible to produce the sheet without the hidden "Common Samples" column.  We could even have the user make the same choices if we used the sample data and just didn't include it.  We just would not be able to catch "conflict resolution conflicts" when the load happens, which could mean confusing debugging to discern what's happened.

It might be possible to reverse engineer to common samples in the peak annotations loader to still be able to handle the scenarios in the examples in the "Purpose" section above, but having the `PeakGroupConflicts` class provide the samples to the `PeakAnnotationsLoader` in a dict like `resolutions[compound][sample] = selected_file` makes the code in the `PeakAnnotationsLoader` very straightforward.

Personally, I feel that having the `Common Samples` column exist, but be hidden unless there's an error is the sweet spot of compromise.  There may exist a less complex means of constructing the sheet.  I have optimized its production and the added running time is negligible in the most common use-case.  But as to whether the column should exist, I think is a question of transparency of what's going on.

The biggest question I see in this discussion is:

**If there are 5 extra samples in the "not selected" file that are not in the selected file, should those PeakGroups get loaded when all the companion samples are loaded from a different file (/scan)?  I.e. Does that cause invalid comparisons in the final data?**

However, the 3-file example is clearly a valid (if unlikely) case, regardless of the answer to the above question.

**Reply by lparsons** [reply]
I've no problem with the "Common Samples" column. I just thought that the example file provided could have more representative data in "Example Samples" so that one could see that it has a subset of the "Common Samples".
